lightning/src/routing/router.rs

```diff
@@ -876,39 +877,85 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
 		// If a caller provided us with last hops, add them to routing targets. Since this happens
 		// earlier than general path finding, they will be somewhat prioritized, although currently
 		// it matters only if the fees are exactly the same.
-		for hop in last_hops.iter() {
+		for route in last_hops.iter() {
+			let first_hop_in_route = &(route.0)[0];
 			let have_hop_src_in_graph =
-				// Only add the last hop to our candidate set if either we have a direct channel or
-				// they are in the regular network graph.
-				first_hop_targets.get(&hop.src_node_id).is_some() ||
-				network.get_nodes().get(&hop.src_node_id).is_some();
+				// Only add the hops in this route to our candidate set if either
+				// we have a direct channel to the first hop or the first hop is
+				// in the regular network graph.
+				first_hop_targets.get(&first_hop_in_route.src_node_id).is_some() ||
+				network.get_nodes().get(&first_hop_in_route.src_node_id).is_some();
 			if have_hop_src_in_graph {
-				// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
-				// really sucks, cause we're gonna need that eventually.
-				let last_hop_htlc_minimum_msat: u64 = match hop.htlc_minimum_msat {
-					Some(htlc_minimum_msat) => htlc_minimum_msat,
-					None => 0
-				};
-				let directional_info = DummyDirectionalChannelInfo {
-					cltv_expiry_delta: hop.cltv_expiry_delta as u32,
-					htlc_minimum_msat: last_hop_htlc_minimum_msat,
-					htlc_maximum_msat: hop.htlc_maximum_msat,
-					fees: hop.fees,
-				};
-				// We assume that the recipient only included route hints for routes which had
-				// sufficient value to route `final_value_msat`. Note that in the case of "0-value"
-				// invoices where the invoice does not specify value this may not be the case, but
-				// better to include the hints than not.
-				if add_entry!(hop.short_channel_id, hop.src_node_id, payee, directional_info, Some((final_value_msat + 999) / 1000), &empty_channel_features, 0, path_value_msat, 0) {
-					// If this hop connects to a node with which we have a direct channel,
-					// ignore the network graph and, if the last hop was added, add our
-					// direct channel to the candidate set.
-					//
-					// Note that we *must* check if the last hop was added as `add_entry`
-					// always assumes that the third argument is a node to which we have a
-					// path.
-					if let Some(&(ref first_hop, ref features, ref outbound_capacity_msat, _)) = first_hop_targets.get(&hop.src_node_id) {
-						add_entry!(first_hop, *our_node_id , hop.src_node_id, dummy_directional_info, Some(outbound_capacity_msat / 1000), features, 0, path_value_msat, 0);
+				// We start building the path from reverse, i.e., from payee
+				// to the first RouteHintHop in the path.
+				let rev_route = route.0.iter().rev().collect::<Vec<_>>();
```

Comment:
Why allocate this as a vec just to index backwards into it? Just access `route.0` in the `prev_hop` calculation, below.

Reply:
I thought it would be easier this way. Would there be a problem with this?

Reply:
`rev_route` is supposed to stay one behind the current iterator value. It would have been much easier to use [prev_iter](https://docs.rs/prev-iter/0.1.2/prev_iter/struct.PrevPeekable.html) but I went without using it. If I zip the two iterators then both of them would essentially give the same value. [`reverse()`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.reverse) might have worked but it does the operation in place which I thought might cause complications.